Implement providing Request in UserRequestError.

Zandor300 opened the pull request:
When an `UserRequestError` is thrown, provide the original `ModbusAbstractRequest` in the error body.

This will allow running createPayload in the main application for debugging purposes.

Currently the only way to get the generated payload is by enabling debug, which can only be done using environment variable:

```
DEBUG=client-request-handler
```

but this will produce a lot of spam of all modbus communication, while I only want to get the payload of only the request I am testing.

Current location where debug for payload is provided:

https://github.com/Cloud-Automation/node-modbus/blob/a7e0d6a247a9771889919d89a0ab8b4c0192ced7/src/client-request-handler.ts#L225

Fixes https://github.com/Cloud-Automation/node-modbus/issues/256

Zandor300 committed:
Implement providing Request in UserRequestError.

Zandor300 committed:
Provide request in error body at additional locations where the request promise can be rejected.

Zandor300 commented:
@stefanpoeter Would you be able to take a look at my changes? Not sure I implemented it correctly or to your liking.

stefanpoeter commented:
Hey @Zandor300 ,
thank you very much for the contribution. The changes are looking good. I think that would be minor change so please increase the minor version in the package.json and reset the patch. Also I noticed that there are no tests for your changes. Would be great if you add some.

Zandor300 committed:
Fix linting issues.

Zandor300 committed:
Provide request in UserRequestError body in additional places.

Zandor300 committed:
Expose ModbusRTURequest and ModbusRTUResponse to match TCP versions which are also exposed.

Zandor300 committed:
Add check for request to some tests to make sure it gets added to the UserRequestError body.

Zandor300 committed:
Remove trailing spaces that my IDE automatically removed.

Zandor300 commented:
@stefanpoeter 
- I added checks to some existing tests that are testing for exception handling. It checks if the request is an instance of `ModbusTCPRequest`/`ModbusRTURequest` for TCP and RTU implementations respectively.

- Also, I only added the request to the `UserRequestError` body in the RTU implementation, I also corrected that.

- Additionally, in the files I edited there were some trailing spaces which I also committed separately to clean up a bit.

- Lastly, I needed to have access to `ModbusRTURequest` in the test so I added an export for that. I think that shouldn't be an issue since the TCP version is also being exposed in the same way.

The PR changes may seem cluttered, but I split up my commits to make it as readable as possible for you.

stefanpoeter commented on src/user-request-error.ts:
```diff
@@ -1,25 +1,30 @@
+  public request?: Req
   public response?: Res
-  constructor ({ err, message, response }: IUserRequestError<Res>) {
+  constructor ({ err, message, request, response }: IUserRequestError<Res, Req>) {
```
Changing the order here is a breaking change. It should be `response, request?` so we stay compatible.

Zandor300 replied:
I've changed the order as requested, but I believe since the parameters are being passed as an object, the order that is being specified here doesn't matter for the enduser.

Also writing the following produces an error:

```js
{ err, message, response, request? }
```

stefanpoeter replied:
Ahh, I am very sorry, I did not notice that they where passed in as a object. Too much C++ lately.

stefanpoeter commented on test/rtu-client-request-handler.test.js:
```diff
@@ -81,8 +82,9 @@ describe('Modbus/RTU Client Request Tests', function () {
-          // Exception type should be ModbusException not crcMismatch or any other 
+          // Exception type should be ModbusException not crcMismatch or any other
           assert.equal(err.err, 'ModbusException')
+          assert.equal(err.request instanceof Modbus.ModbusRTURequest, true)
```
This does not actually test the content of the request. Compare the original request with the one in the error object and not just the requests type. Should be possible to create buffer from both and compare those.

stefanpoeter replied:
There a more of this down the line.

Zandor300 replied:
I've added asserts for testing if the actual object is the same.

I also added 2 additional tests for TCP where this is tested. I had to write these tests differently than the other tests since I can't get access to the actual request body when using something like `tcpClient.readCoils(..)`

stefanpoeter commented:
@Zandor300 Thanks again, overall this looks fine. I added two comments to your changes. And please increase the minor version number in the package.json.

Zandor300 commented:
@stefanpoeter I’m sorry, I can’t find your comments. Did you comment them on the PR changes?

stefanpoeter commented:
Sorry, forgot to submit the review :-)

Zandor300 committed:
Add check if the request bodies provided in UserRequestError match in RTU tests.

Zandor300 committed:
Add new tests to check if the request bodies provided in UserRequestError match in TCP tests.

Zandor300 committed:
Version bump. (v4.0.10)

Zandor300 committed:
Reorder parameters of UserRequestError constructor.

Zandor300 commented:
@stefanpoeter I've made adjustments to my PR. Note my replies to your review comments.

stefanpoeter commented:
As soon as I recover my npm account you'll see the changes there too. Thanks for your contribution. Good work :-)

stefanpoeter commented:
Congrats. Version 4.0.10 is online :-)
